### PR TITLE
PWN-8177 - show refresh fragment when no internet

### PR DESCRIPTION
**Pull request by eduardmaximovich**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-8177

## Description of Work

- added fake refreshing button that checks internet connection state and hides the fragment if connected
- moved ConnectionManager from Ethereum kit to core module
- deprecated `ConnectionManager.Listener` in favor of Kotlin Flow for flexibility
- added method to check if initial tokens have been loaded from json file
- added method to reload initial tokens if not loaded due to internet loss

### CODE REVIEW

**Review by 1sakov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/interactor/RefreshErrorInteractor.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+class RefreshErrorInteractor {
+
+    private val refreshClickFlow = MutableSharedFlow<Unit>(replay = 1)
+
+    fun getRefreshClickFlow() = refreshClickFlow.asSharedFlow()
+
+    fun notifyRefreshClick() {
+        refreshClickFlow.tryEmit(Unit)
+    }
```
[the remark this answers is not in the document]
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/home/interactor/RefreshErrorInteractor.kt:14]: Okay!

File: app/src/main/java/org/p2p/wallet/AppModule.kt [hunk 2]
```diff
@@ -59,6 +60,7 @@ object AppModule {
                 tokenKeyProvider = get()
             )
         }
+        single { ConnectionManager(androidContext()) }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/AppModule.kt:63]: singleOf will work okay too

File: app/src/main/java/org/p2p/wallet/home/HomeModule.kt [hunk 3]
```diff
@@ -40,6 +43,7 @@ object HomeModule : InjectionModule {
     private fun Module.initDataLayer() {
         factory<HomeLocalRepository> { HomeDatabaseRepository(get()) }
         factoryOf(::HomeMapper)
+        factory<RefreshErrorRepository> { RefreshErrorSharedFlowRepository() }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/HomeModule.kt:46]: FactoryOf with bind

File: app/src/main/java/org/p2p/wallet/home/HomeModule.kt [hunk 4]
```diff
@@ -57,6 +61,7 @@ object HomeModule : InjectionModule {
             )
         }
         factoryOf(::SearchInteractor)
+        single { RefreshErrorInteractor(refreshErrorRepository = get()) }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/HomeModule.kt:64]: singleOf

File: app/src/main/java/org/p2p/wallet/home/MainFragment.kt [hunk 5]
```diff
@@ -121,6 +141,24 @@ class MainFragment :
         deeplinksManager.executeTransferPendingAppLink()
     }
 
+    private val refreshErrorFragment: RefreshErrorFragment by unsafeLazy {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/MainFragment.kt:144]: all fields should be placed on top

File: app/src/main/java/org/p2p/wallet/home/interactor/RefreshErrorInteractor.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package org.p2p.wallet.home.interactor
+
+import org.p2p.wallet.home.repository.RefreshErrorRepository
+
+class RefreshErrorInteractor(
+    private val refreshErrorRepository: RefreshErrorRepository
+) {
+    fun getRefreshClickFlow() = refreshErrorRepository.getFlow()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/interactor/RefreshErrorInteractor.kt:8]: explicit return type

File: app/src/main/java/org/p2p/wallet/home/interactor/RefreshErrorInteractor.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+    fun notifyRefreshClick() = refreshErrorRepository.notifyClick()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/interactor/RefreshErrorInteractor.kt:10]: notifyRefreshClicked - past tense, click was done

File: app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package org.p2p.wallet.home.repository
+
+import kotlinx.coroutines.flow.Flow
+
+interface RefreshErrorRepository {
+    fun getFlow(): Flow<Unit>
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt:6]: getFlow of what?

File: app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+    fun notifyClick()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt:7]: Data layer doesn't know anything about clicks and stuff

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt [hunk 10]
```diff
@@ -142,14 +146,12 @@ class HomePresenter(
...
-    private fun attachToPollingTokens() {
-        launch {
-            tokensPolling.shareTokenPollFlowIn(this).filterNotNull().collect { homeState ->
-                state = state.copy(tokens = homeState.solTokens, ethTokens = homeState.ethTokens)
-                initializeActionButtons()
-                handleUserTokensLoaded(homeState.solTokens, homeState.ethTokens)
-                view?.showRefreshing(homeState.isRefreshing)
-            }
+    private suspend fun attachToPollingTokens() {
+        tokensPolling.shareTokenPollFlowIn(this).filterNotNull().collect { homeState ->
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:150]: place each call on new line

File: app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt [hunk 11]
```diff
@@ -161,6 +163,18 @@ class HomePresenter(
         }
     }
 
+    private fun observeInternetConnection() {
+        launch {
+            connectionManager.connectionStatus.collect {
+                if (it) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt:169]: if (it) doesn't give you anything about the check that is done inside if

File: app/src/main/res/layout/fragment_refresh_error.xml [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+    <androidx.constraintlayout.widget.Guideline
+        android:id="@+id/guide_top"
+        android:layout_width="0dp"
+        android:layout_height="0dp"
+        android:orientation="horizontal"
+        app:layout_constraintGuide_percent="0.18" />
+
+    <ImageView
+        android:id="@+id/image_view_box"
```
Comment (gslevinkov) [inline app/src/main/res/layout/fragment_refresh_error.xml:16]: camelCase

File: core/src/main/java/org/p2p/core/network/ConnectionManager.kt [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    private fun notifyHasConnection(hasConnection: Boolean) {
+        listener?.onConnectionChange()
+        _connectionStatus.tryEmit(hasConnection)
+    }
+
+    private fun NetworkCapabilities?.hasValidatedInternet(): Boolean {
+        if (this == null) return false
+
+        return hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+            && hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
```
Comment (gslevinkov) [inline core/src/main/java/org/p2p/core/network/ConnectionManager.kt:52]: I think we leave && on the first line instead of putting it on the new line

**Review by gslevinkov — CHANGES_REQUESTED** [review]

**Comment by gslevinkov** [comment]
This big PR could be split into small ones to review it easier, please do it next time if there is need

File: app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt (hunk 9 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt:7]: What do you mean? I've renamed (locally yet) getFlow -> getClickFlow

File: app/src/main/res/layout/fragment_refresh_error.xml (hunk 12 above)
Comment (eduardmaximovich) [reply app/src/main/res/layout/fragment_refresh_error.xml:16]: https://www.notion.so/0de0d6f69a5c476085552c31b645c61f
item number 6

File: core/src/main/java/org/p2p/core/network/ConnectionManager.kt (hunk 13 above)
Comment (eduardmaximovich) [reply core/src/main/java/org/p2p/core/network/ConnectionManager.kt:52]: 1. It won't fit to 120 symbols in row (static importing leads to implicitness)
2. new line has better readability

File: app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt (hunk 8 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt]: renamed to:
getRefreshClickedFlow()
notifyRefreshClicked()

File: app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt (hunk 9 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt]: In my opinion data layer can't use 'click' or UI-connected terms.  no clicks, only domain words and what it really means. 
we don't need to know about any clicks or other stuff on this level - we are interested in refresh and errors

fun notifyRefreshError - okay
fun getRefreshErrorFlow - okay
or smth else

File: app/src/main/java/org/p2p/wallet/home/HomeModule.kt [hunk 14]
```diff
@@ -43,7 +44,7 @@ object HomeModule : InjectionModule {
     private fun Module.initDataLayer() {
         factory<HomeLocalRepository> { HomeDatabaseRepository(get()) }
         factoryOf(::HomeMapper)
-        factory<RefreshErrorRepository> { RefreshErrorSharedFlowRepository() }
+        factoryOf(::RefreshErrorSharedFlowRepository) bind RefreshErrorRepository::class
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/HomeModule.kt:47]: We name impl of a repository using Remote/Local/InMemory words. 
If it caches values - you can name it RefreshErrorInMemoryRepository: RefreshErrorLocalRepository

File: app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorSharedFlowRepository.kt [hunk 15]
```diff
@@ -7,8 +7,8 @@ class RefreshErrorSharedFlowRepository : RefreshErrorRepository {
 
     private val flow = MutableSharedFlow<Unit>(replay = 1)
 
-    override fun getFlow() = flow.asSharedFlow()
-    override fun notifyClick() {
+    override fun getRefreshClickFlow() = flow.asSharedFlow()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorSharedFlowRepository.kt:10]: explicit return type

File: app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt (hunk 9 above)
Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt]: I see your point, but refreshError suggests that we're notifying about an error. In reality, the logic is about error resolution. So, let's rename 'click' to 'event' instead.

getRefreshEventFlow()
notifyEventRefreshed()

What do you think?

Comment (eduardmaximovich) [reply app/src/main/java/org/p2p/wallet/home/repository/RefreshErrorRepository.kt]: Also the names of `RefreshError[Repository|Interactor]` could probably be changed to something more descriptive if their purpose is not clear.

**Pull request merged**